### PR TITLE
Pseudo Relevance Feedback Interpolation on pysearch

### PR DESCRIPTION
**Pull request by x65han**
Sample Command
----------------
```
python -m pyserini.search \
 -index /Users/johnson/anserini/indexes/lucene-index.robust04.pos+docvectors+raw \
 -topics robust04 \
 -output robust04-prf.txt \
 -prf -alpha 1
```
bm25 + LR on robust04 (alpha=1)
----------------
```
map                   	all	0.2550
P_30                  	all	0.3138
```

### CODE REVIEW

File: pyserini/search/__main__.py [hunk 1]
```diff
@@ -1,33 +1,55 @@
 import re
```
Comment (lintool) [inline pyserini/search/__main__.py:1]: when you get a chance, please add in our standard boilerplate headers. and reranker below.

File: pyserini/search/reranker.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    LR = 'lr'
+    SVM = 'svm'
+
+
+class FusionMethod(enum.Enum):
+    AVG = 'avg'
+
+
+class PseudoRelevanceClassifierReranker:
+    def __init__(self, lucene_index: str, clf_type: List[ClassifierType], r=10, n=100, alpha=1):
```
Comment (lintool) [inline pyserini/search/reranker.py:20]: you probably want alpha=0.5 as a default; defaults should do something reasonable...

File: pyserini/search/reranker.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+from sklearn.linear_model import LogisticRegression
+from sklearn.svm import SVC
+from typing import List
+from ..vectorizer import TfidfVectorizer
+import uuid
+import os
+
+
+class ClassifierType(enum.Enum):
+    LR = 'lr'
```
Comment (lintool) [inline pyserini/search/reranker.py:11]: Maybe we should just give it the full sklearn name?

File: pyserini/search/reranker.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    def set_classifier(self, clf_type: ClassifierType):
+        if clf_type == ClassifierType.LR:
+            self.clf = LogisticRegression()
+        elif clf_type == ClassifierType.SVM:
+            self.clf = SVC(kernel='linear', probability=True)
+        else:
+            raise Exception("Invalid classifier type")
+
+    def get_prf_vectors(self, doc_ids: List[str]):
+        train_docs = doc_ids[:self.r] + doc_ids[-self.n:]
```
Comment (lintool) [inline pyserini/search/reranker.py:40]: Add a comment here to explain what you're doing here... notation might not be familiar to all - and line below.
This is actually a cute use of * notation.

File: pyserini/search/reranker.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    def set_classifier(self, clf_type: ClassifierType):
+        if clf_type == ClassifierType.LR:
+            self.clf = LogisticRegression()
+        elif clf_type == ClassifierType.SVM:
+            self.clf = SVC(kernel='linear', probability=True)
+        else:
+            raise Exception("Invalid classifier type")
+
+    def get_prf_vectors(self, doc_ids: List[str]):
```
Comment (lintool) [inline pyserini/search/reranker.py:39]: this is intended to be private? use _ notation?

File: pyserini/search/reranker.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        return train_vecs, train_labels, test_vecs
+
+    def rerank_with_classifier(self, doc_ids: List[str], search_scores: List[float]):
+        train_vecs, train_labels, test_vecs = self.get_prf_vectors(
+            doc_ids)
+
+        # classification
+        self.clf.fit(train_vecs, train_labels)
+        pred = self.clf.predict_proba(test_vecs)
+        rank_scores = self.normalize([p[1] for p in pred])
```
Comment (lintool) [inline pyserini/search/reranker.py:55]: rank_score -> classifier_scores to be more accurate

File: pyserini/search/reranker.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+        # interpolation
+        interpolated_scores = [a * self.alpha + b * (1-self.alpha)
+                               for a, b in zip(rank_scores, search_scores)]
+
+        return self.sort_dual_list(interpolated_scores, doc_ids)
+
+    def rerank(self, doc_ids: List[str], search_scores: List[float]):
```
Comment (lintool) [inline pyserini/search/reranker.py:63]: I think it'd make sense to take as input an array of hits?
that way, other data (e.g., the document text) can come along for the ride as well (and be used as part of reranking in other reranker implementations).

Comment (x65han) [reply pyserini/search/reranker.py:64]: Ok, I can do that. I was a little hesitant at first, since I don't know how to get the type of hits.

**Pull request merged**